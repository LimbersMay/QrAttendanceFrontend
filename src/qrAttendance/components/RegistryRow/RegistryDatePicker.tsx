import * as React from 'react';
import dayjs, { Dayjs } from 'dayjs';
import Stack from '@mui/material/Stack';
import TextField from '@mui/material/TextField';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { DateTimePicker } from '@mui/x-date-pickers/DateTimePicker';

export const RegistryDatePicker = ({ date, onChangeDate }: { date: string | undefined, onChangeDate: any }) => {
    const [value, setValue] = React.useState<Dayjs | null>(
        dayjs(date),
    );

    const handleChange = (newValue: Dayjs | null) => {
        setValue(newValue);
        onChangeDate(newValue?.toISOString());
    };

    return (
        <LocalizationProvider dateAdapter={AdapterDayjs}>
            <Stack spacing={3}>
                <DateTimePicker
                    label="Date&Time picker"
                    value={value}
                    onChange={handleChange}
                    renderInput={(params) => <TextField {...params} />}
                />
            </Stack>
        </LocalizationProvider>
    );
}