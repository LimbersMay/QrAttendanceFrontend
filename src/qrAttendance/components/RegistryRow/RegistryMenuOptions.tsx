import { MouseEvent, useState } from 'react';
import {Divider, IconButton, Menu, MenuItem, Typography} from "@mui/material";
import {Delete, Edit, MoreVert} from "@mui/icons-material";
import {useUiSlice} from "../../../hooks/useUiSlice";
import {useRegistrySlice} from "../../../hooks/useRegistrySlice";
import {Registry} from "../../interfaces";

const ITEM_HEIGHT = 48;

export const RegistryRowMenuOptions = ({ registry }: { registry: Registry }) => {

    const { openRegistryModal } = useUiSlice();
    const { handleSetActiveRegistry, handleDeleteRegistry } = useRegistrySlice();

    const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
    const open = Boolean(anchorEl);

    const handleClick = (event: MouseEvent<HTMLElement>) => {
        setAnchorEl(event.currentTarget);
        handleSetActiveRegistry(registry);
    };

    const handleClose = () => {
        setAnchorEl(null);
        handleSetActiveRegistry(null);
    };

    return (
        <>
            <IconButton
                aria-label="more"
                aria-controls="long-menu"
                aria-haspopup="true"
                onClick={handleClick}
            >
                <MoreVert />
            </IconButton>
            <Menu
                id="long-menu"
                anchorEl={anchorEl}
                keepMounted
                open={open}
                onClose={handleClose}
                PaperProps={{
                    style: {
                        maxHeight: ITEM_HEIGHT * 4.5,
                        width: '20ch',
                    },
                }}
            >
                <MenuItem onClick={openRegistryModal}>
                    <Edit />
                    <Typography sx={{ml: '7px'}}>Edit</Typography>
                </MenuItem>

                <Divider />

                <MenuItem onClick={handleDeleteRegistry}>
                    <Delete />
                    <Typography sx={{ml: '7px'}}>Delete</Typography>
                </MenuItem>

            </Menu>
        </>
    );
}