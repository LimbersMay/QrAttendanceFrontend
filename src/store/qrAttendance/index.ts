export * from "./group";
export * from "./qrCode";
export * from "./registry";
