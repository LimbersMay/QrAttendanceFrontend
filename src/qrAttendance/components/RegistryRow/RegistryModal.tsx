import React, {ChangeEvent, FormEvent, useEffect, useState} from "react";
import Backdrop from '@mui/material/Backdrop';
import Box from '@mui/material/Box';
import Modal from '@mui/material/Modal';
import Fade from '@mui/material/Fade';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import {Divider, Grid, MenuItem, Select, SelectChangeEvent, TextField} from "@mui/material";
import {useUiSlice} from "../../../hooks/useUiSlice";
import {useRegistrySlice} from "../../../hooks/useRegistrySlice";
import {RegistryDatePicker} from "./RegistryDatePicker";

const style = {
    position: 'absolute' as 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    width: 400,
    bgcolor: 'background.paper',
    border: '2px solid #000',
    boxShadow: 24,
    p: 4,
};

const initialState = {
    name: "",
    checkInTime: "2023-02-18T02:19:01.000Z",
    career: "",
    group: "",
    firstSurname: "",
    secondSurname: "",
}

export const RegistryModal = () => {

    const { active: activeRegistry, updateRegistry } = useRegistrySlice();

    const { isRegistryModalOpen, toggleRegistryModal } = useUiSlice();

    const [formState, setFormState] = useState(activeRegistry || initialState);
    const {name, checkInTime, group, career, firstSurname, secondSurname} = formState;

    useEffect(() => {
        setFormState(activeRegistry || initialState);
    }, [activeRegistry]);

    const changeFormState = ({target}: ChangeEvent<HTMLInputElement> | SelectChangeEvent) => {
        const {name, value} = target;

        setFormState({
            ...formState,
            [name]: value
        })
    }

    const onSelectChange = (event: SelectChangeEvent) => {
        const {name, value} = event.target;

        setFormState({
            ...formState,
            [name]: value
        });
    }

    const onDateChange = (date: string) => {
        setFormState({
            ...formState,
            checkInTime: date
        })
    }

    const onInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        changeFormState(event);
    }

    const onSubmit = (event: FormEvent) => {
        event.preventDefault();

        if (!activeRegistry) return;

        updateRegistry({
            ...activeRegistry,
            ...formState
        });

        console.log(activeRegistry);
        console.log(formState);

        toggleRegistryModal();
    }

    return (
        <Modal
            aria-labelledby="transition-modal-title"
            aria-describedby="transition-modal-description"
            open={isRegistryModalOpen}
            onClose={toggleRegistryModal}
            closeAfterTransition
            BackdropComponent={Backdrop}
            BackdropProps={{
                timeout: 500,
            }}
        >
            <Fade in={isRegistryModalOpen}>
                <Box sx={style}>
                    <form onSubmit={onSubmit}>
                        <Grid container>

                            <Grid item xs={12} sx={{textAlign: "center"}}>
                                <Typography variant="h6">Editing</Typography>
                                <Divider />
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <RegistryDatePicker date={checkInTime} onChangeDate={onDateChange} />
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="Group"
                                    name="group"
                                    value={group}
                                    fullWidth
                                    inputProps={{maxLength: 2, style: {textTransform: 'uppercase'}}}
                                    variant={"outlined"}
                                    placeholder="Group"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <Select fullWidth variant={'outlined'} name="career" value={career} onChange={onSelectChange}>
                                    <MenuItem value="Computer Systems Engineering">Computer Systems Engineering</MenuItem>
                                    <MenuItem value="Civil Engineering">Civil Engineering</MenuItem>
                                    <MenuItem value="Environmental Engineering">Environmental Engineering</MenuItem>
                                    <MenuItem value="Industrial Engineering">Industrial Engineering</MenuItem>
                                    <MenuItem value="Engineering in Administration">Engineering in Administration</MenuItem>
                                    <MenuItem value="Bussines Management Engineering">Bussines Management Engineering</MenuItem>
                                </Select>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="First surname"
                                    name="firstSurname"
                                    value={firstSurname}
                                    fullWidth
                                    variant={"outlined"}
                                    placeholder="firstSurname"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="Second surname"
                                    name="secondSurname"
                                    value={secondSurname}
                                    fullWidth
                                    variant={"outlined"}
                                    placeholder="secondSurname"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="Name"
                                    name="name"
                                    value={name}
                                    fullWidth
                                    variant={"outlined"}
                                    placeholder="name"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <Button
                                    variant="contained"
                                    fullWidth
                                    type={"submit"}
                                >
                                    Save
                                </Button>
                            </Grid>
                        </Grid>
                    </form>
                </Box>
            </Fade>
        </Modal>
    );
}