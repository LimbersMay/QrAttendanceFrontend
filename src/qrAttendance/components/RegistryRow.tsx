import {useState} from "react";

import {
    TableCell,
    IconButton,
    TableRow
} from "@mui/material";
import {Delete, Edit, Save} from "@mui/icons-material";

import {useForm} from "../../hooks/useForm";
import {ConditionalTextField} from "./ConditionalTextField";
import {Registry} from "../interfaces";
import {useAppDispatch} from "../../store";
import {startDeleteRegistry, startUpdateRegistry} from "../../store/qrAttendance";

export const RegistryRow = ({registryRow }: { registryRow: Registry}) => {

    const dispatch = useAppDispatch();

    const [isRowEditing, setIsRowEditing] = useState(false);
    const {formState, onInputChange} = useForm(registryRow);

    const { name, sourname, lastname, date } = formState;

    const handleEdit = () => {
        setIsRowEditing(true);
    }

    const handleSave = () => {
        dispatch(startUpdateRegistry({
            ...registryRow,
            ...formState
        }));
        setIsRowEditing(false);
    }

    const handleDelete = () => {
        dispatch(startDeleteRegistry(registryRow.id));
    }

    return (
        <TableRow>
            <TableCell component="th" scope="row">
                <ConditionalTextField
                    name="date"
                    value={date}
                    onChange={onInputChange}
                    condition={isRowEditing}
                />
            </TableCell>
            <TableCell>
                <ConditionalTextField
                    name="name"
                    value={name}
                    onChange={onInputChange}
                    condition={isRowEditing}
                />
            </TableCell>
            <TableCell align="center">
                <ConditionalTextField
                    name="sourname"
                    value={sourname}
                    onChange={onInputChange}
                    condition={isRowEditing}
                    styles={{width: '100px'}}
                />
            </TableCell>
            <TableCell align="center">
                <ConditionalTextField
                    name="lastname"
                    value={lastname}
                    onChange={onInputChange}
                    condition={isRowEditing}
                    styles={{width: '90px'}}
                />
            </TableCell>
            <TableCell align="center">
                {
                    isRowEditing
                        ? <IconButton onClick={handleSave}><Save/></IconButton>
                        : <IconButton onClick={handleEdit}><Edit /></IconButton>
                }
                <IconButton onClick={handleDelete}>
                    <Delete />
                </IconButton>
            </TableCell>
        </TableRow>
    )
}