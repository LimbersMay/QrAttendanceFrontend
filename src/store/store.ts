import {AnyAction, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {authSlice} from "./auth";
import {groupSlice, qrCodeSlice, registrySlice} from "./qrAttendance";
import socketMiddleware from "./middlewares/socketMiddleware";
export const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        group: groupSlice.reducer,
        qrCode: qrCodeSlice.reducer,
        registry: registrySlice.reducer
    },
    middleware: [socketMiddleware]
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    AnyAction
>
