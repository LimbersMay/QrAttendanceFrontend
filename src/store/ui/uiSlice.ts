import {createSlice} from '@reduxjs/toolkit';
import {RootState} from "../store";

interface uiInitialState {
    isQrCodeModalOpen: boolean;
    isRegistryModalOpen: boolean;
    isShowingQrCode: boolean;
    isTitleModalOpen: boolean;
}

const initialState: uiInitialState = {
    isQrCodeModalOpen: false,
    isRegistryModalOpen: false,
    isShowingQrCode: false,
    isTitleModalOpen: false
}

export const uiSlice = createSlice({
    name: 'ui',
    initialState,
    reducers: {
        onShowQrCode: (state) => {
            state.isShowingQrCode = true;
        },
        onHideQrCode: (state) => {
            state.isShowingQrCode = false;
        },
        onToggleQrCodeModal: (state) => {
            state.isQrCodeModalOpen = !state.isQrCodeModalOpen;
        },
        onToggleRegistryModal: (state) => {
            state.isRegistryModalOpen = !state.isRegistryModalOpen;
        },
        onToggleTitleModal: (state) => {
            state.isTitleModalOpen = !state.isTitleModalOpen;
        }
    }
});


// Action creators are generated for each case reducer function
export const { onToggleQrCodeModal, onShowQrCode, onHideQrCode, onToggleRegistryModal, onToggleTitleModal } = uiSlice.actions;
export const selectUi = (state: RootState) => state.ui;