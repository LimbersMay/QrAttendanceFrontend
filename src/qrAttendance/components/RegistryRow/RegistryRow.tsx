import {
    TableCell,
    TableRow
} from "@mui/material";
import {Registry} from "../../interfaces";
import {RegistryRowMenuOptions} from "./RegistryMenuOptions";
import dayjs from "dayjs";
import {RegistryModal} from "./RegistryModal";

export const RegistryRow = ({registryRow}: { registryRow: Registry }) => {

    const {name, group, firstSurname, secondSurname, checkInTime, career} = registryRow;

    return (
        <TableRow>
            <TableCell component="th" scope="row">
                { dayjs(checkInTime).format('MMMM D, YYYY h:mm A') }
            </TableCell>
            <TableCell align="center">
                { group }
            </TableCell>
            <TableCell align="center">
                { career }
            </TableCell>
            <TableCell align="center" sx={{display: {xs: 'none', sm: 'table-cell'}}}>
                { firstSurname }
            </TableCell>
            <TableCell align="center" sx={{display: {xs: 'none', sm: 'table-cell'}}}>
                { secondSurname }
            </TableCell>
            <TableCell align="center">
                { name }
            </TableCell>
            <TableCell align="center" sx={{display: {xs: 'none', sm: 'table-cell'}}}>
                <RegistryRowMenuOptions
                    registry={registryRow}
                />
            </TableCell>

            <RegistryModal />
        </TableRow>
    )
}