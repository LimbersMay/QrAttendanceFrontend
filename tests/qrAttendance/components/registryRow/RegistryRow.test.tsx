import {Provider} from "react-redux";
import {configureStore} from "@reduxjs/toolkit";
import {render, screen} from "@testing-library/react";
import {uiSlice} from "../../../../src/store/ui/uiSlice";
import {registrySlice} from "../../../../src/store/qrAttendance";
import {initialState} from "../../../fixtures/uiStates";
import {
    withActiveRegistryAndRegistriesState
} from "../../../fixtures/qrAttendanceStates";
import {RegistryRow} from "../../../../src/qrAttendance/components/RegistryRow/RegistryRow";
import {registries} from "../../../fixtures/registryStates";

const store = configureStore({
    reducer: {
        ui: uiSlice.reducer,
        registry: registrySlice.reducer
    },
    preloadedState: {
        ui: {...initialState},
        registry: {...withActiveRegistryAndRegistriesState}
    }
});

describe('tests for <RegistryRow />', () => {
    test('should match the snapshot', () => {
        const {container} = render(
            <Provider store={store}>
                <table>
                    <tbody>
                    <RegistryRow registryRow={registries[0]}/>
                    </tbody>
                </table>
            </Provider>
        );

        expect(container).toMatchSnapshot();
    });

    test('should show the name of the registry', () => {
        render(
            <Provider store={store}>
                <table>
                    <tbody>
                        <RegistryRow registryRow={registries[0]}/>
                    </tbody>
                </table>
            </Provider>
        );

        expect(screen.getByText(registries[0].name)).toBeTruthy();
    });
});
