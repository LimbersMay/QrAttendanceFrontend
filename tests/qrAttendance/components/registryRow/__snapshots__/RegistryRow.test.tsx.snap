// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <RegistryRow /> should match the snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
          scope="row"
        >
          Invalid Date
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ce3qcx-MuiTableCell-root"
        >
          group 1
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ce3qcx-MuiTableCell-root"
        >
          System engineering
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1lu91p5-MuiTableCell-root"
        >
          Limbert
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1lu91p5-MuiTableCell-root"
        >
          May Ek
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ce3qcx-MuiTableCell-root"
        >
          Limbers
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1lu91p5-MuiTableCell-root"
        >
          <button
            aria-controls="long-menu"
            aria-haspopup="true"
            aria-label="more"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="registry-menu-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MoreVertIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
